include/rms/rms.hh

```diff
@@ -1,4 +1,79 @@
+/* daft engineers vocoder project
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
 #ifndef RMS_RMS_HH
 #define RMS_RMS_HH
 
+#include "../audio/audio.hh"
+#include "../pipe.hh"
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <numeric>
+
+namespace rms {
+
+template <std::size_t num_samples> class RMS {
+  private:
+    Pipe<Audio> &input_pipe;
+    Pipe<double> &output_pipe;
+
+    std::array<uint32_t, num_samples> sample_buffer = {0};
+    int sample_buffer_index = 0;
+
+    bool run_thread{};
+    std::thread thread;
+
+  public:
+    RMS(Pipe<Audio> &input_pipe, Pipe<double> &output_pipe) : input_pipe(input_pipe), output_pipe(output_pipe) {
+    }
+
+    void insert(Audio packet) {
+        std::for_each(packet.begin(), packet.end(), [this](const int16_t item) {
+            // this assumes that the 0 point for the signal is max(uint16)/2, precicely in the middle of the range
+
+            uint32_t squared = item * item;
```

Comment:
is there really not a better way of doing a square number?

Reply:
i don't see any issue with this method

Reply:
There's nothing wrong with this, it's clear what's happening, but I was slightly confused. Avoiding floats makes sense.